### PR TITLE
[CR][WIP] Show remaining time left in Eat menu, and add support for showing Weeks

### PR DESCRIPTION
**Pull request by MT-Arnoldussen**
#### Summary
SUMMARY: Interface "Show remaining time left in Eat menu, and add support for showing Weeks"

#### To Do
- [x] Display Time Left in `E`at menu..
- [x] Make Time Left use (relative) rot instead of age.
- [x] ~~Rename "Spoils In" -> "Shelf Life" *everywhere*.~~ Could not find any other place where shelf life is called spoiling.
- [x] Re-add Shelf Life column.
- [x] Add Freshness column.
- [ ] Add Time Left ("Spoils In") to item description.
- [x] Use same cooking/survival requirement as for checking freshness in item description

#### Why
Being able to see how long the food you have will roughly last is very useful. Currently you only have the `(old)` and `(fresh)` tags to work with, but usually most of your food will have neither.

#### What
Instead of the time food would last if newly created, the `E`at menu now shows the amount of time the food has left.

*It will only show the Freshness and Spoils In columns if you have Cooking 3 or Survival 4.* This is consistent with the detailed freshness estimation in the description of food.

* If the remaining time is less than or equal to 6 hours and the item is `(old)` it shows "soon!". This is to prevent having items with "6 hours" left from rotting. (maybe "<6 hours" instead of "soon!"/" 6 hours"  would be better?) 
* If the remaining time is less than or equal to 6 hours and the item is not `(old)`, it shows 6 hours.
* If the remaining time is less than or equal to 1 week, it shows the amount of days, rounded up.
* Otherwise it will round up to the next week.
* This will then be changed into seasons or years if appropriate.

In addition to that, it will now show "weeks" in between "days" and "seasons". This is mostly for people who play with longer season lengths, since seeing something will last 28 or 42 days is kinda weird.

#### Screenshots
Default season length:
short

91 day seasons:
long

Also added this to non-fresh, non-old items if your cooking or survival is too low to see full freshness estimation:
need more skill

#### Potential Issues
With higher amounts of time left the displayed time being rounded up to a week might still be a little too precise.

### CODE REVIEW

**Comment by cainiaowu** [comment]
for unaccurate estimate, add "about" so player dont think it is a bug.

**Comment by nexusmrsep** [comment]
This is a good change, but it should be wider. Some thoughts:

To avoid confusion I'd encourage you to do a similar approach in other places where nominal spoil time in room temperatures is prioritized over current decay level, for example in [e]xamine, where it is shown first, before assessment of actual freshness (5 levels) (added in PR #24057).

Perhaps even re-wording of the nominal spoil time message should be done to accent that its prone to change. Using same words & definitions like "Spoils in ..." that don't mean the same thing in different contexts will lead to confusion and players will instantly mix-up nominal and actual rot times. Those changes should get in in the same time to prevent that.

Optional way to improve Eat menu: a separate column with freshness levels derived from PR #24057 + column with nominal spoil time + column with expected time to rot from now (in any order fitting).
This would be even better in my opinion, because that would allow to weight those values against each other by the player for better understanding.

Another thing: consider linking display of nominal values and actual values. If one in not shown (or approximated) the other should not be show (or approximated too). Why? Because if you show only one, you give the wrong idea to the player. I've seen it in gameplays on YouTube where nominal spoil time was taken for actual by mistake. I should have done it myself in [e]xamine, but I haven't considered it at that time.

**Comment by MT-Arnoldussen** [comment]
>Using same words & definitions like "Spoils in ..." that don't mean the same thing in different contexts will lead to confusion and players will instantly mix-up nominal and actual rot times. 

I didn't think of that, but now you mention it I absolutely agree. I'm thinking of reverting the "Spoils In" column, and adding a "Time Left" column. 

I'm not sure how I would incorporate the relative rot. Ideally you'd be able to sort this menu in various ways, and relative rot should be an option there, but I have looked at sorting this kind of menu before and I have no idea how I would do that.

I'm not sure if a different column for freshness would be ideal. That would mean adding two more columns, plus I'm not sure how often relative freshness is more important than absolute freshness. I actually can't think of a scenario where I would care about relative rot more than actual time left.

**Comment by nexusmrsep** [comment]
> Alternatively maybe "Shelf Life" could replace what is now "Spoils In", and "Spoils In" is the actual time left.

I like the "Shelf life" very much. It's intuitive when stacked with "Spoils in".

Present freshens in examine is actually textified relative rot: it is divided in five tiers by percentage 10|20|20|20|20|10 (sum=100 %), each with a different description given to the player. I abbreviated it to show how it may look like:
- fresh
- quite fresh
- near midlife
- past midlife
- getting older
- old

So lets say you have a food item X with shelf-life of 2 weeks and relative rot 60%. Ideally I'd see:

| FOOD | NUTRITION | QUENCH | JOY | SHELF LIFE | FRESHNESS | SPOILS IN |
| ----- | ------ | -------- | --------- | ------- | ------- | --------- |
| X | 15 | 15 | 15 | 2 weeks | past midlife | 8 days |

With freshness visible (and colored) its easier to discern on first sight why "spoild in" < "shelf life" too.

As for sorting: both freshness and spoils in are `relative rot` derivatives so sorting by `relative rot` will do the trick in both cases.

**Comment by nexusmrsep** [comment]
One more very important thing I noticed:
This: `const time_duration time_left = spoils - age; // estimate. food can be ok if this is <0` can go horribly wrong. If you have food X with 2 weeks shelf life and you check it after a week, when it spent one week frozen in winter time, that your estimate SPOILS IN will be 1 week, and will be far from reality, because freshness will be still: (fresh). `spoils` is constans while `age` is constantly moving forward regardless of true decay. Same with preserved comestibles.
Also: food (in towns etc.) spawns  with non-zero rot to simulate storing conditions and production dates before Cataclysm. Those values can be negative and positive (+/- 20% of item's shelf life). This will not be taken into account with above check.

Consider using `get_relative_rot()` that returns decay in percentage, `get_rot()` that returns `time_duration` of accumulated decay to do this calculation. You can even do something like this: `rng( get_rot() - 1_week, get_rot() + 1_week )` (as `rng()` works well with `time_duration`) to do some randomization if you want/need.

**Comment by MT-Arnoldussen** [comment]
>I like the "Shelf life" very much. It's intuitive when stacked with "Spoils in".

Alright. I will go for a set of columns like you showed first.

>One more very important thing I noticed

Thanks, I completely missed the rot values when I coded this. I just assumed that age slowed down during cold/freezing, and that there was some randomness involved with rotting. Which is clearly not what is happening.

So just to be clear, there is zero randomness involved if you use `get_relative_rot()`? Because if so I could also have the `Time Left` column be more precise than a single day for higher survival skill characters.

**Comment by nexusmrsep** [comment]
Get relative rot is not randomized in any way. It just divides rot by spoil time to get percentage. You can find how it works in map.cpp as far as I remember.

**Comment by nexusmrsep** [comment]
Looks awesome even now.

**Comment by MT-Arnoldussen** [comment]
@narc0tiq I think I did all the changes you listed (even tried having inconsistent spacing around &) and it still doesn't work, although it does fail on a different file so calendar.h should be good now.

How do you get such a detailed report of astyle failings anyway? Would be very useful to have that myself.

**Comment by narc0tiq** [comment]
I ran `make astyle` and copied the `git diff` output. I did have a
different compilation problem after that, but since it's my (fairly
modified) personal branch, I didn't try fixing it, just backed out the
inclusion of this PR.

**Comment by MT-Arnoldussen** [comment]
Alright. I'll have to test these new changes myself, and add this detailed spoils-in information to the item examine screen, and then this PR should be ready.

**Comment by narc0tiq** [comment]
Jenkins, rebuild please.

File: src/item.cpp [hunk 1]
```diff
@@ -970,24 +970,30 @@ std::string item::info(std::vector<iteminfo> &info, const iteminfo_query *parts,
...
+                    info.emplace_back( "DESCRIPTION", _( "* This food looks like it <bad>will be old soon</bad>.  "
                                                          "It's now or never, if you plan to use it." ) );
                 } else if( food_item->is_going_bad() ) {
-                    info.emplace_back( "DESCRIPTION", _( "* This food looks <bad>old</bad>. "
+                    info.emplace_back( "DESCRIPTION", _( "* This food looks <bad>old</bad>.  "
                                                          "It's on a brink of becoming inedible." ) );
+                } else { // unskilled and neither fresh nor old
+                    info.emplace_back( "DESCRIPTION", _( "* This food looks <neutral>fine</neutral>.  "
+                                                         "If you were more skilled in cooking or survival "
+                                                         "you might be able to make a better estimation." ) );
```
Comment (narc0tiq) [inline src/item.cpp]: Having just noticed this while doing a merge (not doing a proper code review, just making a personal branch), I'd like to see this logic set up differently:

```C++
if( !food_item->rotten() ) {
    if( food_item->is_fresh() ) { // even the unskilled can see it's fresh!
        //...
    } else if (food_item->is_going_bad() ) { // even the unskilled can see it's old!
        //...
    } else {
        info.emplace_back( "DESCRIPTION", describe_freshness( rot_progress, skilled) );
    }
```

Where `describe_freshness(double, bool)` can go something like:

```C++
if( !skilled ) {
    return _( "* This food looks <neutral>fine</neutral>.  "
              "If you were more skilled [...]");
} else if( rot_progress >= 0.1 && rot_progress < 0.3 ) {
    return _( "* This food looks <good>still quite fresh</good>.[...]" );
// ...and so on
```

This is just my personal preference, not necessarily project style; please correct me if it seems like I'm suggesting something very wrong (in particular, separating the strings into different locations in the file might be confusing to contributors -- it's usually better to have them all in the same place, and it's nice to see them in order, too).

**Comment by narc0tiq** [comment]
Hey, I've been playing with this for a while and I was getting tired of having to keep fixing the conflicts -- so I just plumbed in a merge commit. Please remember to pull your branch before continuing work on it.

**Comment by MT-Arnoldussen** [comment]
@narc0tiq I kinda stopped working on this due to lack of interest in the game itself and also time issues IRL.

If you want to finish this and/or make your own PR based on your branch feel free to do so, otherwise I will close this PR for now.

**Comment by narc0tiq** [comment]
Ok, I can take care of it in my own PR, then -- you can close/delete the branch if you wish, I have a copy. Thanks for doing the awesome work, I'll make sure it gets finished.

**Comment by narc0tiq** [comment]
Just for clarity, what's the piece you still had left to be done? Adding the time span from the "Spoils In" column to the item description?

**Comment by narc0tiq** [comment]
Ok, I've taken over via #26318 and #26319. Thank you for your effort!

**Pull request closed without merging**